docs/architecture/adr-004-coroutine-effect-system.md

```diff
@@ -0,0 +1,303 @@
+# ADR 004: Coroutine-Based Effect System for Consensus
+
+## Changelog
+
+* 18-03-2025: Initial version
+
+## Status
+
+Accepted
+
+## Context
+
+The Malachite consensus engine needs to interact with its environment (network, storage, cryptography, application logic) at specific points during execution.
+
+Traditional approaches to handling these interactions include:
+
+1. **Callback-based designs** - Provide callback interfaces that must be implemented by users
+2. **Trait-based polymorphism** - Define traits that users implement to provide required functionality
```

Comment:
Is this Rust-specific?

Reply:
The _trait_ nomenclature is, but this is also known as interfaces or subclasses in OOP languages or type classes in FP languages.